libc/src/__support/RPC/rpc.h

```diff
@@ -560,9 +562,9 @@ template <uint16_t opcode> LIBC_INLINE Client::Port Client::open() {
 template <uint32_t lane_size>
 [[clang::convergent]] LIBC_INLINE
     cpp::optional<typename Server<lane_size>::Port>
-    Server<lane_size>::try_open() {
+    Server<lane_size>::try_open(uint32_t start) {
   // Perform a naive linear scan for a port that has a pending request.
-  for (uint32_t index = 0; index < process.port_count; ++index) {
+  for (uint32_t index = start; index < process.port_count; ++index) {
```

Comment:
I wanted to make sure that `handle_server` runs a fixed amount of times because we use the user thread to check it. This prevents the situation where some other kernel's RPC calls perpetually block a user's thread from finishing, e.g.
```
while (!kernel.done)
   server.handle_server() // Could pick up work from other async kernels.
```

I haven't yet introduced sleeping to the busy-wait in OpenMP, which would then probably not want to increase a sleep duration if we just finished a full scan or something.

Reply:
Agreed with Johannes - start until end is weird on a circular buffer. Suggest we check N ports on each call, where N may as well be the size of the buffer

Reply:
> I wanted to make sure that handle_server runs a fixed amount of times because we use the user thread to check it. 

As @JonChesterfield noted, my suggestion does not violate this. Each call should check all ports.

Reply:
That's what I'm intending with this logic. When the user calls `handle_server` it will guaruntee that all `1 through N` ports will be checked. If it finds work on a port, it handles it, then picks back up where it left off. If we make each check go `1 through N` modulo the port size then we'll be in the same situation where we don't know whether or not `handle_server` will run a bounded number of times.

Reply:
So, I am not seeing this to run through all N ports. Let's see the last picked one is N-2, then index is N-1, and this checks only a single port, namely N-1. Before, it always checked N ports, but always starting at 0. What we want is N (=all) ports are checked, but starting at the last picked index so you can't starve a port since the one we picked last is checked last in the next run. My proposal was to run N checks, nothing unbounded.

Reply:
Look at the greater scope. This isn't `handle_server` this is `try_open`. `handle_server` passes in `start` as the last position it found a port. If we instead always check starting at `N` we will then continuously check forever as long as there is work still being submitted faster to the RPC interface than it can consume it. We can potentially do that, but I was just thinking that it would be good to guarantee that this routine, when called by a blocking thread, will return in a specific number of checks.

Reply:
Ok. I think I see the disconnect. Your scheme doesn't actually provide the guarantee I thought I was, which is why we talk past each other. In your patch, one call to handle_server will always start to process the ports from 0 till the end. With this patch it will ensure termination as it checks each port once. That is clearly better than the code before, not necessarily because it could get stuck, but because it could starve ports. I don't think stuck performing work is too bad for the host thread. For now I don't see what the benefit would be to return to the caller of handle_server, just to let them call it again.

That all said, what I thought you were doing is to remember the index of the last port that was handled across calls. Such that you'd scan from there the next time. Now, you still always prefer port 0 over port 5 even if you just handled port 0 last time. That said, this patch at least ensures both are handled during a handle_server call. If index is persistent though, which is what I thought, port 5 would be prioritized over port 0 if port 0 was handled last time. In that scheme, you need to iterate from 0 to index at the end, since index did not start at 0. Hence my request for the extra iterations.

I still think the persistent index has benefits when it comes to fairness, but I agree this patch in itself does what you describe in a sensible way.

Reply:
> For now I don't see what the benefit would be to return to the caller of handle_server, just to let them call it again.

My main concern was if we had the user's thread running the server it could potentially be blocked for a long time if it keeps picking up some other kernel's work. I could instead do the method where we start at the last known index, but  always check N ports.